daemon/restart.go

```diff
@@ -35,11 +36,25 @@ func (daemon *Daemon) containerRestart(container *container.Container, seconds i
 		defer daemon.Unmount(container)
 	}
 
-	if err := daemon.containerStop(container, seconds); err != nil {
+	// set AutoRemove flag to false before stop so the container won't be
+	// removed during restart process
+	autoRemove := container.HostConfig.AutoRemove
```

Comment:
This is a little different. We need this to ensure `docker restart` a `AutoRemove` container can succeed but not stop+remove it. 
So it's by design.